File: leetcode/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/cpp/main.cpp
#include <cstdio>
#include <cstdlib>
#include <string>
#include <unordered_map> // Use unordered_map for better performance
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x, TreeNode* left, TreeNode* right)
        : val(x)
        , left(left)
        , right(right)
    {
    }
};

class Solution {
public:
    // Recursive function to build the tree
    TreeNode* buildTreeRec(vector<int>& preorder, int& preorder_idx, unordered_map<int, int>& inorder_map, int inorder_start, int inorder_end)
    {
        // Base case: if there are no elements to construct the subtree
        if (inorder_start > inorder_end) {
            return nullptr;
        }

        // Get the root value from the current preorder index
        int root_val   = preorder[preorder_idx];
        TreeNode* root = new TreeNode(root_val);

        // Increment preorder index for the next call
        preorder_idx++;

        // If there is only one element, no need to recurse further
        if (inorder_start == inorder_end) {
            return root;
        }

        // Find the index of the root in inorder traversal
        int inorder_root_idx = inorder_map[root_val];

        // Build left subtree
        root->left = buildTreeRec(preorder, preorder_idx, inorder_map, inorder_start, inorder_root_idx - 1);

        // Build right subtree
        root->right = buildTreeRec(preorder, preorder_idx, inorder_map, inorder_root_idx + 1, inorder_end);

        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
    {
        // Check if preorder and inorder are valid
        if (preorder.empty() || inorder.empty() || preorder.size() != inorder.size()) {
            return nullptr;
        }

        // Create a map to store value -> index relations for inorder traversal
        unordered_map<int, int> inorder_map;
        for (int i = 0; i < inorder.size(); i++) {
            inorder_map[inorder[i]] = i;
        }

        int preorder_idx = 0; // Start from the first element in preorder
        return buildTreeRec(preorder, preorder_idx, inorder_map, 0, inorder.size() - 1);
    }
};

// Function to generate a string representation of the tree for printing
string getPrintTreeStr(TreeNode* root, string prefix, bool isLeft)
{
    if (!root) {
        return "";
    }
    string str = "";
    str += prefix;
    str += (isLeft) ? "├──" : "└──";
    str += to_string(root->val);
    str += "\n";
    string next_prefix = prefix;
    next_prefix += (isLeft) ? "│   " : "    ";
    if (root->left) {
        str += getPrintTreeStr(root->left, next_prefix, true);
    }
    if (root->right) {
        str += getPrintTreeStr(root->right, next_prefix, false);
    }
    return str;
}

// Function to print the tree
void printTree(TreeNode* root)
{
    printf("%s", getPrintTreeStr(root, "", false).c_str());
}

int main(int argc, char* argv[])
{
    Solution s {};

    // Test Case 1
    vector<int> preorder1 { 3, 9, 20, 15, 7 };
    vector<int> inorder1 { 9, 3, 15, 20, 7 };
    TreeNode* root1 = s.buildTree(preorder1, inorder1);
    printf("Test Case 1:\n");
    printTree(root1);
    printf("\n");

    // Test Case 2
    vector<int> preorder2 { 1, 2 };
    vector<int> inorder2 { 1, 2 };
    TreeNode* root2 = s.buildTree(preorder2, inorder2);
    printf("Test Case 2:\n");
    printTree(root2);
    printf("\n");

    // Test Case 3 (Invalid: Duplicate values in inorder)
    // Uncommenting the following lines will cause incorrect behavior due to duplicate '1's
    /*
    vector<int> preorder3 { 1, 2, 3 };
    vector<int> inorder3  { 3, 1, 1 };
    TreeNode* root3 = s.buildTree(preorder3, inorder3);
    printf("Test Case 3:\n");
    printTree(root3);
    printf("\n");
    */

    // Test Case 4
    vector<int> preorder4 { 3, 2, 1, 4 };
    vector<int> inorder4 { 1, 2, 3, 4 };
    TreeNode* root4 = s.buildTree(preorder4, inorder4);
    printf("Test Case 4:\n");
    printTree(root4);
    printf("\n");

    return 0;
}

File: leetcode/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/rust/src/main.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
struct Solution;
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

fn print_tree(root: Option<Rc<RefCell<TreeNode>>>) {
    fn print_tree_rec(root: &Option<Rc<RefCell<TreeNode>>>, level: i32) {
        if let Some(node) = root {
            print_tree_rec(&node.borrow().right, level + 1);
            for _ in 0..level {
                print!("    ");
            }
            println!("{}", node.borrow().val);
            print_tree_rec(&node.borrow().left, level + 1);
        }
    }
    print_tree_rec(&root, 0);
}

impl Solution {
    fn build_tree_rec(
        preorder: &Vec<i32>,
        inorder: &Vec<i32>,
        pre_start: i32,
        in_start: i32,
        in_end: i32,
        inorder_map: &HashMap<i32, i32>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if pre_start >= preorder.len() as i32 || in_start > in_end {
            return None;
        }
        let root_val = preorder[pre_start as usize];
        let root = Rc::new(RefCell::new(TreeNode::new(root_val)));
        let in_root_index = inorder_map[&root_val];
        let left_size = in_root_index - in_start;
        let right_size = in_end - in_root_index;
        let left = Solution::build_tree_rec(
            preorder,
            inorder,
            pre_start + 1,
            in_start,
            in_root_index - 1,
            inorder_map,
        );
        let right = Solution::build_tree_rec(
            preorder,
            inorder,
            pre_start + left_size + 1,
            in_root_index + 1,
            in_end,
            inorder_map,
        );
        root.borrow_mut().left = left;
        root.borrow_mut().right = right;
        Some(root)
    }
    pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut inorder_map = HashMap::new();
        for (i, &val) in inorder.iter().enumerate() {
            inorder_map.insert(val, i as i32);
        }
        return Solution::build_tree_rec(
            &preorder,
            &inorder,
            0,
            0,
            inorder.len() as i32 - 1,
            &inorder_map,
        );
    }
}

fn main() {
    let preorder = vec![3, 9, 20, 15, 7];
    let inorder = vec![9, 3, 15, 20, 7];
    let root = Solution::build_tree(preorder, inorder);
    print_tree(root);

    let preorder = vec![-1];
    let inorder = vec![-1];
    let root = Solution::build_tree(preorder, inorder);
    print_tree(root);

    let preorder = vec![1, 2, 3];
    let inorder = vec![3, 2, 1];
    let root = Solution::build_tree(preorder, inorder);
    print_tree(root);

    let preorder = vec![1, 2, 3];
    let inorder = vec![1, 2, 3];
    let root = Solution::build_tree(preorder, inorder);
    print_tree(root);
}
